File: bytebrew-srv/internal/domain/pairing_token.go
package domain

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"math/big"
	"time"
)

const (
	// pairingTokenBytes is the number of random bytes for the full token (256-bit)
	pairingTokenBytes = 32
	// pairingTokenExpiry is how long a pairing token remains valid
	pairingTokenExpiry = 5 * time.Minute
	// shortCodeMax is the upper bound for generating a 6-digit code
	shortCodeMax = 1000000
)

// PairingToken represents a temporary token for mobile device pairing
type PairingToken struct {
	Token           string // Full token (256-bit, hex encoded)
	ShortCode       string // 6-digit code for manual entry
	ExpiresAt       time.Time
	Used            bool
	ServerID        string // Unique server identifier
	ServerPublicKey []byte // X25519 public key (sent to mobile)
	ServerPrivateKey []byte // X25519 private key (kept until pairing completes)
}

// NewPairingToken creates a new PairingToken with a random token, 6-digit short code, and 5 min expiry
func NewPairingToken(serverID string) (*PairingToken, error) {
	if serverID == "" {
		return nil, fmt.Errorf("server_id is required")
	}

	tokenBytes := make([]byte, pairingTokenBytes)
	if _, err := rand.Read(tokenBytes); err != nil {
		return nil, fmt.Errorf("generate token: %w", err)
	}
	token := hex.EncodeToString(tokenBytes)

	codeNum, err := rand.Int(rand.Reader, big.NewInt(shortCodeMax))
	if err != nil {
		return nil, fmt.Errorf("generate short code: %w", err)
	}
	shortCode := fmt.Sprintf("%06d", codeNum.Int64())

	return &PairingToken{
		Token:     token,
		ShortCode: shortCode,
		ExpiresAt: time.Now().Add(pairingTokenExpiry),
		Used:      false,
		ServerID:  serverID,
	}, nil
}

// IsExpired returns true if the token has expired
func (t *PairingToken) IsExpired() bool {
	return time.Now().After(t.ExpiresAt)
}

// IsValid returns true if the token is not expired and not used
func (t *PairingToken) IsValid() bool {
	return !t.IsExpired() && !t.Used
}

// MarkUsed marks the token as used
func (t *PairingToken) MarkUsed() {
	t.Used = true
}

File: bytebrew-cli/src/domain/entities/PairingToken.ts
/**
 * PairingToken represents a temporary token for mobile device pairing.
 * Port from Go: bytebrew-srv/internal/domain/pairing_token.go
 *
 * Pure domain entity — no external dependencies.
 */

/** Default pairing token expiry: 5 minutes */
export const PAIRING_TOKEN_EXPIRY_MS = 15 * 60 * 1000;

export interface PairingTokenProps {
  readonly token: string;
  readonly shortCode: string;
  readonly expiresAt: Date;
  readonly used: boolean;
  readonly serverPublicKey: Uint8Array;
  readonly serverPrivateKey: Uint8Array;
}

export class PairingToken {
  readonly token: string;
  readonly shortCode: string;
  readonly expiresAt: Date;
  readonly used: boolean;
  readonly serverPublicKey: Uint8Array;
  readonly serverPrivateKey: Uint8Array;

  private constructor(props: PairingTokenProps) {
    this.token = props.token;
    this.shortCode = props.shortCode;
    this.expiresAt = props.expiresAt;
    this.used = props.used;
    this.serverPublicKey = props.serverPublicKey;
    this.serverPrivateKey = props.serverPrivateKey;
  }

  /**
   * Creates a new PairingToken with the given token and short code.
   * Expiry is set to 5 minutes from now.
   *
   * Crypto key generation is NOT done here (domain stays pure).
   * Use withKeys() to attach keys after generation in infrastructure layer.
   */
  static create(token: string, shortCode: string): PairingToken {
    if (!token) throw new Error('token is required');
    if (!shortCode) throw new Error('short_code is required');

    return new PairingToken({
      token,
      shortCode,
      expiresAt: new Date(Date.now() + PAIRING_TOKEN_EXPIRY_MS),
      used: false,
      serverPublicKey: new Uint8Array(0),
      serverPrivateKey: new Uint8Array(0),
    });
  }

  /**
   * Restores a PairingToken from persisted data (no validation).
   */
  static fromProps(props: PairingTokenProps): PairingToken {
    return new PairingToken(props);
  }

  /**
   * Returns a new PairingToken with server X25519 keys attached.
   */
  withKeys(serverPublicKey: Uint8Array, serverPrivateKey: Uint8Array): PairingToken {
    return new PairingToken({
      token: this.token,
      shortCode: this.shortCode,
      expiresAt: this.expiresAt,
      used: this.used,
      serverPublicKey,
      serverPrivateKey,
    });
  }

  /**
   * Returns true if the token has expired.
   */
  isExpired(): boolean {
    return Date.now() > this.expiresAt.getTime();
  }

  /**
   * Returns true if the token is not expired and not used.
   */
  isValid(): boolean {
    return !this.isExpired() && !this.used;
  }

  /**
   * Returns a new PairingToken marked as used.
   */
  markUsed(): PairingToken {
    return new PairingToken({
      token: this.token,
      shortCode: this.shortCode,
      expiresAt: this.expiresAt,
      used: true,
      serverPublicKey: this.serverPublicKey,
      serverPrivateKey: this.serverPrivateKey,
    });
  }
}
